src/Features/Core/Portable/Snippets/AbstractSnippetService.cs

```diff
@@ -46,10 +49,23 @@ public ISnippetProvider GetSnippetProvider(string snippetIdentifier)
         /// </summary>
         public async Task<ImmutableArray<SnippetData>> GetSnippetsAsync(Document document, int position, CancellationToken cancellationToken)
         {
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            if (syntaxFacts.IsInNonUserCode(syntaxTree, position, cancellationToken))
+            {
+                return ImmutableArray<SnippetData>.Empty;
+            }
+
+            var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
+            var context = document.GetRequiredLanguageService<ISyntaxContextService>().CreateContext(document, semanticModel, position, cancellationToken);
```

Comment:
Instead of multiple providers creating `SyntaxContext`s, now it's pre-computed here and used for all of them.

It also avoids the code duplication of getting `semanticModel` and `context`